### PR TITLE
[Codegen][Common] Reorder layout analysis to preserve anchored to_layout

### PR DESCRIPTION
**Pull request by raikonenfnu**
TL;DR: Re-ordered forward and backward layout + added support to propagate through arith.select S.T setting to_layout at certain locations in decomposition gets honored. 

This change makes it easier to honor to_layout conversions generated by the users and/or higher level decompositons. Since intuitively to_layout op has the effect that ops that come after it shall take/be based on it's layout. Loading the forward propagation first, also encourage priortize ops to take layouts from anchors / to_layout op that comes before it. 

The motivation behind this PR was to solve the issue where despite us emitting to_layout op at certain locations, we ended up generating redundant to_simd earlier in the graph. This results us to not have control on where the layout conversion actually happens.

One example of this case is when we emit a to_layout conversion for FP8 Attention. We emit a to_layout right before the 2nd contract and right after the truncf to FP8. Since we do enforcement(backward propagation) first and then propagate (forward propagation), we ended up generating two layout conversions, one the original one we emitted, the second one determined by layout analysis and placed in the middle of a elemwise op that is part of the softmax.

This forces us to do shuffles/layout resolution in FP32 which hurts the performance. Additionally, this is also not where we(compiler writers) intended the layout conversion to happen.

Additionally, we made modification to getAgreedLayout to handle elemwise propagation where some operands are not expected to have layout such as arith.select who's condition operand probably don't have layouts.

### CODE REVIEW

**Comment by raikonenfnu** [comment]
I think with this change it's easier to honor to_layout conversions generated by the users and/or higher level decompositons. Since intuitively to_layout op has the effect that ops that come after it shall take/be based on it's layout. Loading the forward propagation first, also encourage priortize ops to take layouts from anchors / to_layout op that comes before it. 

Tests that changed seems to changed for the better.

For example one of test:

```mlir
%read = vector.transfer_read %src
%layout_conv_0 = to_layout %read to #layout_0
%add = arith.addf %layout_conv_0, %layout_conv_0
%layout_conv_1 = to_layout %add to #layout_1
vector.transfer_write %layout_conv_1, %dst
``` 

Used to generate something like:

```mlir
%read = vector.transfer_read %src [#layout_0]
%layout_conv_0 = to_layout %read from #layout_0 to #layout_1 
%add = arith.addf %layout_conv_0, %layout_conv_0 [#layout_1]
%layout_conv_1 = to_layout %add to #layout_1
vector.transfer_write %layout_conv_1, %dst

// Note that read seems to be prematurely converted to layout1 and add is happening in layout1
// while the original IR seem to intend add to be happening in layout0. 
``` 

Which now generates:
```mlir
%read = vector.transfer_read %src [#layout_0] 
%add = arith.addf %read, %read [#layout_0] 
%layout_conv_1 = to_layout %add from #layout_0  to #layout_1
vector.transfer_write %layout_conv_1, %dst [#layout1]

// Now add and read is in layout 0 and only gets converted to layout 1 after add, 
// which seems more aligned with the  intend of the original IR.
```

File: compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp [hunk 1]
```diff
@@ -899,6 +910,13 @@ DistributionLayout *PropagateLayout::getLatticeElement(Value val) {
 /// ==========================================================================
 
 LogicalResult EnforceLayout::initialize(Operation *root) {
+  // Set layout for anchor ops or resolve if need be.
+  for (auto [val, layout] : anchors) {
+    DistributionLayout *latticeEl = getLatticeElement(val);
+    ChangeResult changed = latticeEl->resolve(layout);
+    propagateIfChanged(latticeEl, changed);
+  }
+
```
Comment (Groverkss) [inline compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp]: Why is this needed? We already have a transfer function for layout operations, right?

File: compiler/src/iree/compiler/Codegen/Common/test/vector_layout_analysis.mlir [hunk 2]
```diff
@@ -188,6 +188,63 @@ builtin.module attributes { transform.with_named_sequence } {
...
+
+    func.return %contract_1 : vector<64x64xf16>
+  }
+
+  transform.named_sequence @__transform_main(%variant_op: !transform.any_op {transform.readonly}) {
+    %top_level_func = transform.structured.match ops{["func.func"]} in %variant_op : (!transform.any_op) -> !transform.any_op
+    transform.iree.test_vector_layout_analysis %top_level_func : !transform.any_op
+    transform.yield
+  }
+}
```
Comment (Groverkss) [inline compiler/src/iree/compiler/Codegen/Common/test/vector_layout_analysis.mlir]: Can we have a smaller test? A test with 2 contracts is big imo.

File: compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp [hunk 3]
```diff
@@ -899,6 +910,13 @@ DistributionLayout *PropagateLayout::getLatticeElement(Value val) {
 /// ==========================================================================
 
 LogicalResult EnforceLayout::initialize(Operation *root) {
+  // Set layout for anchor ops or resolve if need be.
+  for (auto [val, layout] : anchors) {
+    DistributionLayout *latticeEl = getLatticeElement(val);
+    ChangeResult changed = latticeEl->resolve(layout);
+    propagateIfChanged(latticeEl, changed);
+  }
```
Comment (Groverkss) [inline compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp]: Can we instead look for all to_layout operations in the root op and initialize using them?

**Review by Groverkss — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/Common/test/vector_layout_analysis.mlir (hunk 2 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Common/test/vector_layout_analysis.mlir]: Yeah I thought would be good to make a small repro of how attention looks like, but I can just make it two "anchored: adds lol

Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Common/test/vector_layout_analysis.mlir]: Made the repro more minimal, this repro is a proxy to simulate the issue we had where a redundant to_layout conversion was emitted by the analysis that forced us to shuffle in F32 prematurely (as opposed to shuffling in F8) btw.

File: compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp (hunk 3 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp]: Makes sense, thanks :)

**Review by Groverkss — APPROVED** [review]
LGTM, Thanks!
